### PR TITLE
Adds engine connection handling and refactors C++ driver

### PR DESCRIPTION
**Pull request by janash**

### CODE REVIEW

**Comment by janash** [comment]
@taylor-a-barnes - initialization function updated -ready for review!

File: {{cookiecutter.repo_name}}/templates/cpp/{{cookiecutter.repo_name}}.cpp [hunk 1]
```diff
@@ -2,18 +2,44 @@
...
+
+    engines[engine_name.get()] = comm;
+
+  }
+
+  return engines;
+
+}
+
+void initialization(int *argc, char ***argv, MPI_Comm *mpi_world_comm) {
```
Comment (taylor-a-barnes) [inline {{cookiecutter.repo_name}}/templates/cpp/{{cookiecutter.repo_name}}.cpp]: For clarity, I would recommend renaming these to `argc_ptr`, `argv_ptr`, and `mpi_world_comm_ptr`, respectively.

Comment (taylor-a-barnes) [inline {{cookiecutter.repo_name}}/templates/cpp/{{cookiecutter.repo_name}}.cpp:35]: Hi Jessica.  For clarity, I would suggest changing the names of these arguments to `argc_ptr`, `argv_ptr`, and `mpi_world_comm_ptr`, respectively.

**Comment by taylor-a-barnes** [comment]
Thanks very much for these additions.  I just have the small comment above.

File: {{cookiecutter.repo_name}}/templates/cpp/{{cookiecutter.repo_name}}.cpp (hunk 1 above)
Comment (janash) [reply {{cookiecutter.repo_name}}/templates/cpp/{{cookiecutter.repo_name}}.cpp]: These have been changed.

**Comment by taylor-a-barnes** [comment]
Thanks very much!

**Pull request merged**